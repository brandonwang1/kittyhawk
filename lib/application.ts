import { Construct } from 'constructs';
import { Deployment, DeploymentProps } from './deployment';
import { Ingress, IngressProps } from './ingress';
import { Service, ServiceProps } from './service'
import { Certificate } from './certificate';

/**
 * Warning: Before editing any interfaces, make sure that none of the interfaces will have 
 * property names that conflict with each other. Typescript may not throw an error and it 
 * could cause problems.
 */
export interface ApplicationProps extends IngressProps, DeploymentProps,
  ServiceProps { }


export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${appname}`

    new Service(this, fullname, props);

    new Deployment(this, fullname, props);

    if (props.ingress) {
      new Ingress(this, fullname, props)

      new Certificate(this, fullname, props)
    }
  }
}


export interface DjangoApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * Override extraEnv from ContainerOptions to make it mutable.
   */
  extraEnv?: { name: string, value: string }[];

  /**
     * A list of host rules used to configure the Ingress.
     *
     * @default undefined
     */
  ingress?: { host: string, paths: string[] }[];
}

export class DjangoApplication extends Application {
  constructor(scope: Construct, appname: string, props: DjangoApplicationProps) {

    // Add the domain as an env variable.
    props.extraEnv = props.extraEnv || [];
    props.extraEnv.push({ name: 'DOMAIN', value: props.domain });

    // Configure the ingress using ingressPaths.
    props.ingress = [{ host: props.domain, paths: props.ingressPaths }]


    // Check if the env variables contains DOMAIN
    const envDomain = props.extraEnv?.filter(env => (env.name === 'DOMAIN'));
    if (envDomain?.length > 1) {
      throw new Error('Django Application should not redefine a DOMAIN enviroment variable.')
    }

    // Check if the env variables contains DJANGO_SETTINGS_MODULE
    const envSettingsModule = props.extraEnv?.filter(env => (env.name === 'DJANGO_SETTINGS_MODULE'));
    if (!envSettingsModule?.length) {
      throw new Error('Django Application must define a DJANGO_SETTINGS_MODULE enviroment variable.')
    }

    // If everything passes, construct the Application.
    super(scope, appname, props);
  }
}

export interface ReactApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * Override extraEnv from ContainerProps to make it mutable.
   */
  extraEnv?: { name: string, value: string }[];

  /**
     * A list of host rules used to configure the Ingress.
     *
     * @default undefined
     */
  ingress?: { host: string, paths: string[] }[];
}

export class ReactApplication extends Application {
  constructor(scope: Construct, appname: string, props: ReactApplicationProps) {

    // Add the domain as an env variable.
    props.extraEnv = props.extraEnv || [];
    props.extraEnv.push({ name: 'DOMAIN', value: props.domain });

    // Configure the ingress using ingressPaths.
    props.ingress = [{ host: props.domain, paths: props.ingressPaths }]

    // Check if the env variables contains DOMAIN
    const envDomain = props.extraEnv?.filter(env => (env.name === 'DOMAIN'));
    if (envDomain?.length > 1) {
      throw new Error('React Application should not redefine a DOMAIN enviroment variable.')
    }

    // Check if the env variables contains PORT
    const envSettingsModule = props.extraEnv?.filter(env => (env.name === 'PORT'));
    if (!envSettingsModule?.length) {
      throw new Error('React Application must define a PORT enviroment variable.')
    }

    // If everything passes, construct the Application.
    super(scope, appname, props);
  }
}