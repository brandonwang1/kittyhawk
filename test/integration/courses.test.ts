import { Construct } from 'constructs';
import { Application } from '../../src/application'
import { chartTest } from '../utils'

export function buildCoursesChart(scope: Construct) {

  /** Penn Courses (Part of it) **/
  new Application(scope, 'backend', {
    image: 'pennlabs/penn-courses-backend',
    secret: 'penn-courses',
    cmd: ['celery', 'worker', '-A', 'PennCourses', '-Q', 'alerts,celery', '-linfo'],
    replicas: 3,
    extraEnv: [{ name: 'PORT', value: '80' },
      { name: 'DJANGO_SETTINGS_MODULE', value: 'PennCourses.settings.production' }],
    ingress: [
      { host: 'penncourseplan.com', paths: ['/api', '/admin', '/accounts', '/assets'] },
      { host: 'penncoursealert.com', paths: ['/api', '/admin', '/accounts', '/assets', '/webhook'] },
      { host: 'review.penncourses.org', paths: ['/api', '/admin', '/accounts', '/assets'] },
    ],
  })
}

test('Penn Courses', () => chartTest(buildCoursesChart));
