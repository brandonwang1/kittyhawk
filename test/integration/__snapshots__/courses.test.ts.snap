// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Penn Courses 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-backend",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-backend",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-backend",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-backend",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 3,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-backend",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-backend",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "command": Array [
                "celery",
                "worker",
                "-A",
                "PennCourses",
                "-Q",
                "alerts,celery",
                "-linfo",
              ],
              "env": Array [
                Object {
                  "name": "PORT",
                  "value": "80",
                },
                Object {
                  "name": "DJANGO_SETTINGS_MODULE",
                  "value": "PennCourses.settings.production",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "penn-courses",
                  },
                },
              ],
              "image": "pennlabs/penn-courses-backend:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-backend",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "penncourseplan.com",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/api",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/admin",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/accounts",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/assets",
              },
            ],
          },
        },
        Object {
          "host": "penncoursealert.com",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/api",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/admin",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/accounts",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/assets",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/webhook",
              },
            ],
          },
        },
        Object {
          "host": "review.penncourses.org",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/api",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/admin",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/accounts",
              },
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-backend",
                  "servicePort": 80,
                },
                "path": "/assets",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "penncourseplan.com",
          ],
          "secretName": "penncourseplan-com-tls",
        },
        Object {
          "hosts": Array [
            "penncoursealert.com",
          ],
          "secretName": "penncoursealert-com-tls",
        },
        Object {
          "hosts": Array [
            "review.penncourses.org",
          ],
          "secretName": "penncourses-org-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "penncourseplan-com",
    },
    "spec": Object {
      "dnsNames": Array [
        "penncourseplan.com",
        "*.penncourseplan.com",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "penncourseplan-com-tls",
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "penncoursealert-com",
    },
    "spec": Object {
      "dnsNames": Array [
        "penncoursealert.com",
        "*.penncoursealert.com",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "penncoursealert-com-tls",
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "penncourses-org",
    },
    "spec": Object {
      "dnsNames": Array [
        "penncourses.org",
        "*.penncourses.org",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "penncourses-org-tls",
    },
  },
]
`;
